modules/ting_search/ting_search.module

```diff
@@ -827,41 +851,91 @@ function ting_search_sort_form($form_state) {
-    'date_ascending' => t('Date (Ascending)'),
-    'date_descending' => t('Date (Descending)'),
-  );
-
-  // Do not show sort-options specific for rank_frequency if
-  // custom_ranking is enabled or another rank is selected.
-  $default_sort = variable_get('ting_sort_default', 'rank_frequency');
-  $custom_ranking = variable_get('ting_ranking_custom', FALSE);
-  if ($default_sort == 'rank_frequency' && empty($custom_ranking)) {
-    $more_options = array(
-      'rank_title' => t('Best match on title'),
-      'rank_creator' => t('Best match on creator'),
-      'rank_subject' => t('Best match on subject'),
-    );
-    $options += $more_options;
+function ting_search_sort_key_from_sort(TingSearchSort $sort) {
+  foreach (ting_search_sort_options() as $key => $info) {
+    if ($info['sort'] === $sort) {
+      return $key;
+    }
   }
+  return NULL;
+}
+
+/**
+ * Produce a list of sort option suitable for a form options element.
+ *
+ * @return array
+ *   machine-name -> label lisl
+ */
+function ting_search_sort_form_options() {
+  $options_full = ting_search_sort_options();
+  array_walk($options_full, function (&$value, $key) {
+    return $value = $value['label'];
+  });
 
   // Add label to the front of the options.
-  foreach ($options as $key => $option) {
-    $options[$key] = t('Sort by: !sort', array('!sort' => $option));
+  foreach ($options_full as $key => $option) {
+    $options_full[$key] = t('Sort by: !sort', array('!sort' => $option));
   }
 
+  return $options_full;
+}
+
+/**
+ * Set default sort options.
+ *
+ * @return array
+ *   Returns an array of sort options.
+ */
+function ting_search_sort_options() {
+  $options = [
+    '' => [
+      'label' => t('Ranking'),
+      'sort' => NULL,
+    ],
+    'title_ascending' => [
+      'label' => t('Title (Ascending)'),
+      'sort' => new TingSearchSort(TingSearchCommonFields::TITLE, TingSearchSort::DIRECTION_ASCENDING),
+    ],
+    'title_descending' => [
+      'label' => t('Title (Descending)'),
+      'sort' => new TingSearchSort(TingSearchCommonFields::TITLE, TingSearchSort::DIRECTION_DESCENDING),
+    ],
+    'creator_ascending' => [
+      'label' => t('Creator (Ascending)'),
+      'sort' => new TingSearchSort(TingSearchCommonFields::AUTHOR, TingSearchSort::DIRECTION_ASCENDING),
+    ],
+    'creator_descending' => [
+      'label' => t('Creator (Descending)'),
+      'sort' => new TingSearchSort(TingSearchCommonFields::AUTHOR, TingSearchSort::DIRECTION_DESCENDING),
+    ],
+    'date_ascending' => [
+      'label' => t('Date (Ascending)'),
+      'sort' => new TingSearchSort(TingSearchCommonFields::DATE, TingSearchSort::DIRECTION_ASCENDING),
+    ],
+    'date_descending' => [
+      'label' => t('Date (Descending)'),
+      'sort' => new TingSearchSort(TingSearchCommonFields::DATE, TingSearchSort::DIRECTION_ASCENDING),
+    ],
+  ];
+
+  // Add provider sorts.
+  $provider_sorts = ding_provider_invoke('search', 'sort_options');
```

Comment:
Should this be optional? E.g. the provider does not have to provide additional sort options.

Reply:
it sure should!